### ISSUE TITLE
Question about light direction estimation

### ISSUE THREAD

**Issue by g3956** [mworchel/differentiable-shadow-mapping#2]
Hi, thanks for your interesting work.

I went through your paper and found that you can also perform light direction estimation task with differentiable shadow mapping.
However, I do not find the corresponding code on this GitHub page.

I simply made the direction differentiable and put it in the optimizer but I got the error "Trying to backward through the graph a second time".  Could you please provide me the training code as reported in Table 9? Any suggestion is helpful.

**Comment by mworchel** [comment]
Hi,

do you mind sharing your modifications? Which example did you modify? 

For the light pose estimation, I simply set up a direction vector with `requires_grad = True`, and in each iteration use it to create a `DirectionalLight` instance that is passed to the `SimpleRenderer.render`. Also, don't forget to scale the direction vector to length 1 in each iteration.

**Comment by g3956** [comment]
Thank you for your help. I know how to fix the issue. 
The point is to use it to create a `DirectionalLight` in **each iteration**.

If I create the `DirectionalLight` as a global variation, the issue occurs.

**Comment by mworchel** [comment]
Hm, interesting. Still, the issue should not occur. Did you setup the optimizer to optimize `DirectionalLight.direction`?

**Comment by g3956** [comment]
In fact, I create a `direction` with `nn.Paramters` Class and setup a corresponding optimizer like
`direction = torch.tensor([[0.57, 0.57, 0.57]], dtype=torch.float32, device=device, requires_grad=True)` and 
`optimizer = torch.optim.Adam([direction], lr=0.3)`.

In each iteration, if I create light as `light = DirectionalLight(direction=F.normalize(direction))` in the loop, it is OK, but if I create light before the training begins, the issue occurs.

**Comment by g3956** [comment]
Anyway, the issue is solved, thanks again for your interesting work and help.

**Comment by mworchel** [comment]
Glad it helped and you found a working solution.

I re-investigated your issue and I am able to directly optimize `DirectionalLight.direction` without any issues, for example using the following script:
```python
import nvdiffrast.torch as dr
import matplotlib.pyplot as plt
import torch

from diffshadow import *
from diffshadow.simple_renderer import *

device         = torch.device('cuda:0')
render_context = dr.RasterizeGLContext(device=device)

# Set up renderer and the scene (camera, meshes)
renderer = SimpleRenderer(render_context)

plane = create_plane_mesh([0, 0, 0], [0, 1, 0], size=1, device=device)

camera = Camera(
    view_matrix=create_lookat_matrix([0., 0.9, 3], [0, 0, 0], [0, 1, 0], device=device),
    projection_matrix=create_perspective_projection_matrix(60, device=device)
)

# Define a reference light direction and render the target image
direction_ref = torch.tensor([[0.1, 0.01, 0.0]], dtype=torch.float32, device=device)
direction_ref = torch.nn.functional.normalize(direction_ref, dim=-1)
light_ref     = DirectionalLight(direction_ref)

img_ref = renderer.render(meshes=[plane], lights=[light_ref], camera=camera)

# Define the optimized light direction and render the initial image
direction_opt = torch.tensor([[0.57, 0.57, 0.57]], dtype=torch.float32, device=device)
direction_opt = torch.nn.functional.normalize(direction_opt, dim=-1)
light_opt     = DirectionalLight(direction_opt) 

img_initial = renderer.render(meshes=[plane], lights=[light_opt], camera=camera)

# Directly optimize DirectionalLight.direction
light_opt.direction.requires_grad_(True)        
optimizer = torch.optim.Adam([light_opt.direction], lr=0.01)

losses = []
for _ in range(100):
    img_opt = renderer.render(meshes=[plane], lights=[light_opt], camera=camera)
    loss = torch.mean((img_opt - img_ref)**2)

    losses.append(float(loss.item()))

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    with torch.no_grad():
        light_opt.direction.data = torch.nn.functional.normalize(light_opt.direction.data, dim=-1)

fig, axs = plt.subplots(1, 4, figsize=(16, 4))
axs[0].imshow(to_display_image(img_ref, grayscale_to_rgb=True))
axs[0].set_title("Reference")
axs[1].imshow(to_display_image(img_initial, grayscale_to_rgb=True))
axs[1].set_title("Initial")
axs[2].imshow(to_display_image(img_opt, grayscale_to_rgb=True))
axs[2].set_title("Optimized")
axs[3].plot(losses)

plt.show()
```

The script produces the following output
![grafik](https://github.com/mworchel/differentiable-shadow-mapping/assets/9449513/28a55fc9-e6e7-4e49-9358-ca36eda05274)
